logger/drain/simple/drain.go

```diff
@@ -0,0 +1,130 @@
+// there are two compelling reasons to redial periodically:
+//
+//  1. We don't want DNS changes on the remote end of the drain to go unnoticed for too long.
+//
+//  2. If the drain is using TCP, the underlying TCP stack can potentially take a very long time
+//     waiting for acks and retrying send for packets that haven't been acked.  This creates a
+//     large window where packets can be spewed into the ether (without any warning) before the
+//     problem is detected.  By redialing periodically, we create the opportunity for a failed TCP
+//     handshake-- which tells us sooner that something is wrong.
+//
+// For efficiency we want the refresh interval to be high.  For resiliency, we want it to be low.
+// One minute has been arbitrarily selected as a sensible balance of these two concerns.
+const connRefreshInterval = 1 * time.Minute
+
+// This determines how many failed dial attempts are required before the drain is muted.
+const maxFailedConns = 5
+
+// This determines how much time we're willing to spend dialing.
+const dialTimeout = 10 * time.Second
+
+// This is how long the drain is muted for after repeated connection failures.
+const mutePeriod = 5 * time.Minute
+
+type logDrain struct {
+	proto string
+	uri   string
+	conn  net.Conn
+	muted bool
+	mutex sync.Mutex
+}
+
+// NewDrain returns a pointer to a new instance of a drain.LogDrain
+func NewDrain(drainURL string) (*logDrain, error) {
+	u, err := url.Parse(drainURL)
+	if err != nil {
+		return nil, err
+	}
+	var proto string
+	if u.Scheme == "udp" || u.Scheme == "syslog" {
+		proto = "udp"
+	} else if u.Scheme == "tcp" {
+		proto = "tcp"
+	} else {
+		return nil, fmt.Errorf("Invalid drain url scheme: %s", u.Scheme)
+	}
+	return &logDrain{proto: proto, uri: u.Host + u.Path}, nil
+}
+
+// Send forwards the provided log message to an external destination
+func (d *logDrain) Send(message string) error {
+	if d.muted {
+		return nil
+	}
+	d.mutex.Lock()
+	defer d.mutex.Unlock()
+	conn, err := d.getConnection(false)
+	if err != nil {
+		return err
+	}
+	_, err = fmt.Fprintln(conn, message)
+	if err != nil {
+		// Try again with a new connection in case the issue was a broken pipe
+		conn, err = d.getConnection(true)
+		if err != nil {
+			return err
+		}
+		_, err = fmt.Fprintln(conn, message)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// getConnection returns a usable connection, often without needing to redial, but still
+// redialing when advised.
+func (d *logDrain) getConnection(forceNew bool) (net.Conn, error) {
+	// If we have a connection, it's not old, and we're not focing a new one...
+	if d.conn != nil && !forceNew {
+		// then return the existing connection
+		return d.conn, nil
+	}
+	// If ANY of those conditions weren't met, it's time for a new connection.
+	// If we have an existing one, close it and nil it out, too for good measure.
+	if d.conn != nil {
+		d.conn.Close()
```

Comment:
some connection types return an error if we fail to properly close it. We should log that error if necessary